File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int size(TreeNode* root, int count){
        if(root == nullptr){
            return count;
        }
        else{
            return (size(root->left, 0) + size(root->right, 0) + 1);
        }
    }
    int kthSmallest(TreeNode* root, int k) {
        int l = size(root->left, 0);
        // cout << k << " " << l << "\n";
        if(k > l+1){
            // cout << "case 1" << "\n";
            return kthSmallest(root->right, k-l-1);
        }
        else if(k < l+1){
            // cout << "case 1" << "\n";
            return kthSmallest(root->left, k);
        }
        else{
            return root->val;
        }
    }
};

File: 230-kth-smallest-element-in-a-bst/230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def numberOfNodes(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        return (1 + self.numberOfNodes(root.left) +  self.numberOfNodes(root.right))
    
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        leftHeight = self.numberOfNodes(root.left)
        
        if leftHeight + 1 == k:
            return root.val
        
        if leftHeight + 1 < k:
            return self.kthSmallest(root.right, k - leftHeight - 1)
        
        
        return self.kthSmallest(root.left, k)